crux/src/Crux/Goal.hs

```diff
@@ -389,48 +392,81 @@ proveGoalsOnline bak opts _ctxt explainFailure (Just gs0) =
 
   hasUnsatCores = unsatCores opts && not (yicesMCSat opts)
 
+  howManyAbducts = case (getNAbducts opts) of Just n  -> n
+                                              Nothing -> 0
+
   failfast = proofGoalsFailFast opts
 
   go (start,end) sp assumptionsInScope gn gs nameMap = do
+    -- traverse goal tree
     case gs of
-
+      -- case: assumption in context for all the contained goals
       Assuming asms gs1 ->
         do ps <- flattenAssumptions sym asms
            forM_ ps $ \asm ->
              unless (trivialAssumption asm) $
+               -- extract predicate from assumption
                do let p = assumptionPred asm
+                  -- create formula, assert to SMT solver, create new name and add to nameMap
                   nm <- doAssume =<< mkFormula conn p
                   bindName nm (Left asm) nameMap
+           -- recursive call
            res <- go (start,end) sp (assumptionsInScope <> asms) gn gs1 nameMap
            return (Assuming (mconcat (map singleAssumption ps)) res)
-
+      -- case: proof obligation in the context of all previously-made assumptions
       Prove p ->
+        -- number of processed goals gives goal number to prove
         do goalNumber <- totalProcessedGoals <$> readIORef gn
            start goalNumber
-           nm <- doAssume =<< mkFormula conn =<< notPred sym (p ^. labeledPred)
+           -- negate goal, create formula
+           t <- mkFormula conn =<< notPred sym (p ^. labeledPred)
+           -- assert formula to SMT solver, create new name and add to nameMap
+           -- This is done in a new assertion frame if abduction is turned on
+           if (howManyAbducts /= 0) then
```

Comment:
Code nit pick: the parentheses around this expression are redundant:

```hs
           if howManyAbducts /= 0 then
```

Similarly for other occurrences of `if (...) ...` in this module.